### PR TITLE
[Prototype] OCI Manifest Reference Support

### LINKED ISSUE #6: Support the referrer field on image manifest
See #5

Something needs to change in here: https://github.com/oci-playground/distribution/blob/main/registry/handlers/manifests.go#L290

cc @akashsinghal

### PR DESCRIPTION
**Pull request by akashsinghal**
- based on https://github.com/oci-playground/distribution/tree/may-17
- updates `ocischemamanifest` handler to index referrers and new link file in storage on PUT
- updates `getReferrers` to support `ArtifactManifest` and `OCIManifest` types
- updates `getReferrers` to set `Content-Type` header in response to OCI index mediatype

Limitations:
- manifest delete path will not clean up extra link files added
- no GC for Artifact references or OCI manifest references
- no tests added

Closes #6

### CODE REVIEW

**Comment by akashsinghal** [comment]
Here are some sample commands to test
```
# checkout this branch https://github.com/akashsinghal/distribution/tree/addReferenceSupport

# run the updated distribution
make

bin/registry serve cmd/registry/config-example-with-extensions.yml

# push some image to registry
podman pull alpine

podman image tag alpine:latest localhost:5000/oci-playground/alpine:latest

podman push --tls-verify=false localhost:5000/oci-playground/alpine:latest

# push OCI Manifest with Reference and list artifacts (https://github.com/oci-playground/regclient/blob/7b7297b0eed91d52575a7210628a1b43e847705b/README.md)
bin/regctl registry set --tls disabled localhost:5000

echo "test artifact" | bin/regctl artifact put --refers localhost:5000/oci-playground/alpine:latest \
  --annotation org.opencontainers.artifact.type=test \
  --annotation org.example.greeting=hello \
  --annotation org.example.location=world

bin/regctl artifact list localhost:5000/oci-playground/alpine:latest --format '{{ jsonPretty .}}' | jq
```

**Comment by akashsinghal** [comment]
Here's some sample commands to test
```
# checkout this branch https://github.com/akashsinghal/distribution/tree/addReferenceSupport

# run the updated distribution
make

bin/registry serve cmd/registry/config-example-with-extensions.yml

# push some image to registry
podman pull alpine

podman image tag alpine:latest localhost:5000/oci-playground/alpine:latest

podman push --tls-verify=false localhost:5000/oci-playground/alpine:latest

# push OCI Manifest with Reference and list artifacts (https://github.com/oci-playground/regclient/blob/7b7297b0eed91d52575a7210628a1b43e847705b/README.md)
bin/regctl registry set --tls disabled localhost:5000

echo "test artifact" | bin/regctl artifact put --refers localhost:5000/oci-playground/alpine:latest \
  --annotation org.opencontainers.artifact.type=test \
  --annotation org.example.greeting=hello \
  --annotation org.example.location=world

bin/regctl artifact list localhost:5000/oci-playground/alpine:latest --format '{{ jsonPretty .}}' | jq
```

File: registry/handlers/blobupload.go [hunk 1]
```diff
@@ -147,15 +146,20 @@ func (buh *blobUploadHandler) PatchBlobData(w http.ResponseWriter, r *http.Reque
...
+				if err != nil {
+					buh.Errors = append(buh.Errors, errcode.ErrorCodeUnknown.WithDetail(err.Error()))
+					return
+				}
+
+				if clInt != (end-start)+1 {
+					buh.Errors = append(buh.Errors, v2.ErrorCodeSizeInvalid)
+					return
+				}
+		*/
```
Comment (jdolitsky) [inline registry/handlers/blobupload.go:162]: wondering if this still needed? issue w/ regctl only?

Comment (sajayantony) [reply registry/handlers/blobupload.go:162]: @sudo-bmitch  Were you able to identify if this was an bug on the distribution side?

**Review by jdolitsky — APPROVED** [review]
once again, in favor of merging considering this is a fork. we can always resolve outstanding issues in the fork before real submission

File: registry/handlers/blobupload.go (hunk 1 above)
Comment (sudo-bmitch) [reply registry/handlers/blobupload.go:162]: Regctl was fixed. Off by one on my range end pointer.

**Pull request merged**